calypso/csadmin/main.go

```diff
@@ -83,3 +91,234 @@ func authorize(c *cli.Context) error {
+			chainConfig.BlockInterval*10, nil)
+		if err != nil {
+			return nil, errors.New("couldn't get proof: " + err.Error())
+		}
+		exist, err := proof.InclusionProof.Exists(iid)
+		if err != nil {
+			return nil, errors.New("error while checking if proof exist: " + err.Error())
+		}
+		if !exist {
+			return nil, errors.New("proof not found")
+		}
+		match := proof.InclusionProof.Match(iid)
+		if !match {
+			return nil, errors.New("proof does not match")
+		}
+
+		return proof, nil
+	}
+
+	writeProof, err := getProof("writeid")
+	if err != nil {
+		return errors.New("failed to get write proof: " + err.Error())
+	}
+	readProof, err := getProof("readid")
+	if err != nil {
+		return errors.New("failed to get read proof: " + err.Error())
+	}
+
+	decryptKey := &calypso.DecryptKey{Write: *writeProof, Read: *readProof}
+
+	reply := &calypso.DecryptKeyReply{}
+	oc := onet.NewClient(cothority.Suite, calypso.ServiceName)
+	err = oc.SendProtobuf(cl.Roster.List[0], decryptKey, reply)
+	if err != nil {
+		return errors.New("failed to send protobuf decryptkey: " + err.Error())
+	}
+
+	if c.Bool("export") {
+		// In case the --export option is provided, the DecryptKeyReply is
+		// encoded and sent to STDOUT.
+		buf, err := protobuf.Encode(reply)
+		if err != nil {
+			return errors.New("failed to encode reply: " + err.Error())
+		}
+		reader := bytes.NewReader(buf)
+		_, err = io.Copy(os.Stdout, reader)
+		if err != nil {
+			return errors.New("failed to copy to stdout: " + err.Error())
+		}
+		return nil
+	}
+
+	fmt.Fprintf(c.App.Writer, "Got decrypt reply:\n"+
+		"- C: %s\n"+
+		"- xHat: %s\n"+
+		"- X: %s\n", reply.C, reply.XhatEnc, reply.X)
+
+	return nil
+}
+
+// recover decrypts an encrypted key stored in a DecryptKeyReply. It expects the
+// DecryptKeyReply to be protobuf encoded and passed in STDIN.
+// With the --export option, the recovered data is sent to STDOUT.
+func recover(c *cli.Context) error {
+	decryptKeyReplyBuf, err := ioutil.ReadAll(os.Stdin)
+	if err != nil {
+		return errors.New("failed to read from stding: " + err.Error())
+	}
+
+	dkr := calypso.DecryptKeyReply{}
+	err = protobuf.Decode(decryptKeyReplyBuf, &dkr)
+	if err != nil {
+		return errors.New("failed to decode decryptKeyReply: " + err.Error())
+	}
+
+	bcArg := c.String("bc")
+	if bcArg == "" {
+		return errors.New("--bc flag is required")
+	}
+
+	cfg, _, err := lib.LoadConfig(bcArg)
+	if err != nil {
+		return err
+	}
+
+	keyPath := c.String("key")
+	var signer *darc.Signer
+	if keyPath == "" {
+		signer, err = lib.LoadKey(cfg.AdminIdentity)
+	} else {
+		signer, err = lib.LoadSigner(keyPath)
+	}
+	if err != nil {
+		return errors.New("failed to load key file: " + err.Error())
+	}
+
+	xc, err := signer.GetPrivate()
+	if err != nil {
+		return errors.New("failed to get private key: " + err.Error())
+	}
+
+	xcInv := xc.Clone().Neg(xc)
+	XhatDec := dkr.X.Clone().Mul(xcInv, dkr.X)
+	Xhat := XhatDec.Clone().Add(dkr.XhatEnc, XhatDec)
+	XhatInv := Xhat.Clone().Neg(Xhat)
```

Comment:
is it not possible to use the `RecoverKey` function?

Reply:
absolutely!